delfin/drivers/dell_emc/power_store/rest_handler.py

```diff
@@ -0,0 +1,958 @@
+        return list_volume
+
+    def get_volume_used_capacity(self, volume_id, volume_type, used_capacity):
+        if volume_type == constants.VolumeType.THICK:
+            return used_capacity
+        data = {'entity': consts.SPACE_METRICS_BY_VOLUME,
+                'entity_id': volume_id}
+        volumes_spaces = self.rest_call(self.REST_GENERATE_URL, data, 'POST')
+        if volumes_spaces:
+            volumes_space = \
+                volumes_spaces[consts.DigitalConstant.MINUS_ONE_INT]
+            used_capacity = volumes_space.get('logical_used')
+        return used_capacity
+
+    def get_disks(self, storage_id):
+        disk_list = []
+        hardware_list = self.rest_call(self.REST_HARDWARE_URL)
+        for hardware in hardware_list:
+            lifecycle_state = hardware.get('lifecycle_state')
+            if consts.CHARACTER_DRIVE != hardware.get('type') or \
+                    lifecycle_state == consts.CHARACTER_EMPTY:
+                continue
+            extra_details = hardware.get('extra_details')
+            capacity = ''
+            firmware = ''
+            physical_type = constants.DiskPhysicalType.UNKNOWN
+            if extra_details:
+                firmware = extra_details.get('firmware_version')
+                drive_type = extra_details.get('drive_type')
+                if drive_type in consts.DiskType.ALL:
+                    continue
+                physical_type = consts.DISK_PHYSICAL_TYPE.get(
+                    drive_type, constants.DiskPhysicalType.UNKNOWN)
+                capacity = extra_details.get('size')
+            disk_result = {
+                'name': hardware.get('name'),
+                'storage_id': storage_id,
+                'native_disk_id': hardware.get('id'),
+                'serial_number': hardware.get('serial_number'),
+                'manufacturer': 'DELL EMC',
+                'firmware': firmware,
+                'capacity': capacity,
+                'status': consts.DISK_STATUS_MAP.get(
+                    lifecycle_state, constants.DiskStatus.NORMAL),
+                'physical_type': physical_type,
+                'logical_type': constants.DiskLogicalType.UNKNOWN,
+                'location': str(hardware.get('slot'))
+            }
+            disk_list.append(disk_result)
+        return disk_list
+
+    def get_controllers(self, storage_id):
+        list_controllers = []
+        nodes = self.get_node()
+        ips = self.get_ip()
+        hardware_list = self.rest_call(self.REST_HARDWARE_URL)
+        for hardware in hardware_list:
+            lifecycle_state = hardware.get('lifecycle_state')
+            if consts.CHARACTER_NODE != hardware.get('type') or \
+                    lifecycle_state == consts.CHARACTER_EMPTY:
+                continue
+            slot = hardware.get('slot')
+            appliance_id = hardware.get('appliance_id')
+            node_id = nodes.get('{}{}'.format(appliance_id, slot), {})
```

Comment:
Recommended using `f'{appliance_id}{slot}'` , so as other places.

Reply:
It has been modified.

node_id = nodes.get(f'{appliance_id}{slot}')
 address = ips.get(f'{appliance_id}{node_id}')